spirv_glsl.cpp

```diff
@@ -3348,100 +3305,285 @@ void CompilerGLSL::emit_extension_workarounds()
+				const std::string name = Supp::getExtensionName(e);
+
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_shuffle:
+					for (const char *t : TYPES)
+						statement(string(t) + " subgroupBroadcastFirst(" + t +
+						          " value) { return shuffleNV(value, 0, gl_WarpSizeNV); }");
+					for (const char *t : TYPES)
+						statement(string(t) + " subgroupBroadcast(" + t +
+						          " value, uint id) { return shuffleNV(value, id, gl_WarpSizeNV); }");
+					break;
+				case Supp::ARB_shader_ballot:
+					for (const char *t : TYPES)
+						statement(string(t) + " subgroupBroadcastFirst(" + t +
+						          " value) { return readInvocationARB(value); }");
+					for (const char *t : TYPES)
+						statement(string(t) + " subgroupBroadcast(" + t +
+						          " value, uint id) { return readInvocationARB(value, id); }");
+					break;
+				}
+			}
+			statement("#endif");
 		}
-		statement("#undef SUBGROUP_ALL_EQUAL_WORKAROUND");
+		else if (shader_subgroup_supporter.isFeatureRequested(Supp::SubgroupBallotFindLSB_MSB))
+		{
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(Supp::SubgroupBallotFindLSB_MSB, result);
 
-		statement("#elif defined(GL_NV_gpu_shader5)");
-		statement("bool subgroupAll(bool value) { return allThreadsNV(value); }");
-		statement("bool subgroupAny(bool value) { return anyThreadNV(value); }");
-		statement("bool subgroupAllEqual(bool value) { return allThreadsEqualNV(value); }");
+			for (uint32_t i = 0u; i < exts.size(); ++i)
+			{
+				const Supp::Candidate e = exts[i];
+				const std::string name = Supp::getExtensionName(e);
 
-		statement("#elif defined(GL_AMD_gcn_shader)&&defined(GL_AMD_gpu_shader_int64)");
-		statement("bool subgroupAll(bool value) { return ballotAMD(value)==ballotAMD(true); }");
-		statement("bool subgroupAny(bool value) { return ballotAMD(value)!=0ull; }");
-		statement("bool subgroupAllEqual(bool value) { uint64_t b=ballotAMD(value); return b==0uLL || "
-		          "b==ballotAMD(true); }");
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_group:
+					statement("uint subgroupBallotFindLSB(uvec4 value) { return findLSB(value.x); }");
+					statement("uint subgroupBallotFindMSB(uvec4 value) { return findMSB(value.x); }");
+					break;
+				}
+			}
+			statement("#else");
+			statement("uint subgroupBallotFindLSB(uvec4 value) {");
+			statement("int firstLive = findLSB(value.x);");
+			statement("return uint(firstLive != -1 ? firstLive : (findLSB(value.y) + 32));");
+			statement("}");
+			statement("uint subgroupBallotFindMSB(uvec4 value) {");
+			statement("int firstLive = findMSB(value.y);");
+			statement("return uint(firstLive != -1 ? (firstLive + 32) : findMSB(value.x));");
+			statement("}");
+			statement("#endif");
+		}
+		else if (shader_subgroup_supporter.isFeatureRequested(Supp::SubgroupAll_Any_AllEqual))
```

Comment:
split this feature into 2, one for bool and one for other types.

The other type feature will need to imply the usage of bool feature  + subgroup broadcast first)